examples/pytorch/rgcn-hetero/entity_classify.py

```diff
@@ -0,0 +1,392 @@
+                 dropout=0.0):
+        super(RelGraphConvHeteroEmbed, self).__init__()
+        self.embed_size = embed_size
+        self.g = g
+        self.bias = bias
+        self.activation = activation
+        self.self_loop = self_loop
+
+        # create weight embeddings for each node for each relation
+        self.embeds = nn.ParameterList()
+        for srctype, etype, dsttype in g.canonical_etypes:
+            embed = nn.Parameter(th.Tensor(g.number_of_nodes(srctype), self.embed_size))
+            nn.init.xavier_uniform_(embed, gain=nn.init.calculate_gain('relu'))
+            self.embeds.append(embed)
+
+        # bias
+        if self.bias:
+            self.h_bias = nn.Parameter(th.Tensor(embed_size))
+            nn.init.zeros_(self.h_bias)
+
+        # weight for self loop
+        if self.self_loop:
+            self.self_embeds = nn.ParameterList()
+            for ntype in g.ntypes:
+                embed = nn.Parameter(th.Tensor(g.number_of_nodes(ntype), embed_size))
+                nn.init.xavier_uniform_(embed,
+                                        gain=nn.init.calculate_gain('relu'))
+                self.self_embeds.append(embed)
+
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self):
+        """ Forward computation
+
+        Returns
+        -------
+        torch.Tensor
+            New node features.
+        """
+        g = self.g.local_var()
+        funcs = {}
+        for i, (srctype, etype, dsttype) in enumerate(g.canonical_etypes):
+            g.nodes[srctype].data['embed-%d' % i] = self.embeds[i]
+            funcs[(srctype, etype, dsttype)] = (fn.copy_u('embed-%d' % i, 'm'), fn.mean('m', 'h'))
+        g.multi_update_all(funcs, 'sum')
+        
+        hs = [g.nodes[ntype].data['h'] for ntype in g.ntypes]
+        for i in range(len(hs)):
+            h = hs[i]
+            # apply bias and activation
+            if self.self_loop:
+                h = h + self.self_embeds[i]
+            if self.bias:
+                h = h + self.h_bias
+            if self.activation:
+                h = self.activation(h)
+            h = self.dropout(h)
+            hs[i] = h
+        return hs
+
+class EntityClassify(nn.Module):
+    def __init__(self,
+                 g,
+                 h_dim, out_dim,
+                 num_bases,
+                 num_hidden_layers=1,
+                 dropout=0,
+                 use_self_loop=False):
+        super(EntityClassify, self).__init__()
+        self.g = g
+        self.h_dim = h_dim
+        self.out_dim = out_dim
+        self.rel_names = list(set(g.etypes))
+        self.num_bases = None if num_bases < 0 else num_bases
+        self.num_hidden_layers = num_hidden_layers
+        self.dropout = dropout
+        self.use_self_loop = use_self_loop
+
+        self.embed_layer = RelGraphConvHeteroEmbed(
+            self.h_dim, g, activation=F.relu, self_loop=self.use_self_loop,
+            dropout=self.dropout)
```

Comment:
I think `RelGraphConvHeteroEmbed` can be replaced with `RelGraphConvHetero` whose inputs are also learnable parameters:

```python
def __init__(self, ...):
    # snip...
    self.embed_layer = RelGraphConvHeteroEmbed(...)
    self.embeds = nn.ParameterDict({
        ntype: nn.Parameter(torch.randn(g.number_of_nodes(ntype), h_dim))
        for ntype in g.ntypes})
    # snip...
def forward(self, ...):
    h = self.embeds
    for layer in self.layers:
        h = layer(self.g, h)
    return h
```

Reply:
I've tried that. It's different with the original modeling. The original RGCN uses W_r * H, and if H is an one-hot encoding, W_r (total shape: R x [N, D]) is equal to a learnable node embedding w.r.t. to relation r. Your modeling is equal to W_r * W_v (shape: R x [D, D] and [N, D]), which in my understanding, is a regularization/re-parameterization of the original W_r. Although it makes sense to me a lot, I found I need use a different hyper-parameter setting to achieve the original result.

Reply:
I see.  It is actually using a different embedding for each relation.  Missed that.  My bad.